### ISSUE TITLE
improve overall experience

### ISSUE THREAD

**Issue by awetzel** [awetzel/neovim-elixir#4]
Hop, text below is a copy of @dm1try proposal on #1 to continue the thread on a new issue.

---

I agree, the integration testing is important here. Also some functional tests for the main modules can help new developers. I mostly use [`espec` framework](https://github.com/antonmi/espec). So I can try to help but if I can use it for this :)

> For the logging thing, yes, I think a 10 LOC File Backend for Logger would be nice with comments on the confix.exs for debugging.
> Then I think it would be nice to have one branch of vim-elixir containing the Host per Erlang version to ensure compatibility with your version.

:+1: 

> About a different project for host and client I am not sure exactly what you mean since the only addition to make the client a host are :

Ok, thanks for the clarifying this. I've noticed the "smell" - `MIX_ENV=host` it seems like the env var is used for make the difference between apps but not the difference between environments. 
I thought that client responsibilities: connect to nvim,retrieve the nvim API/provide an interface to call this API; host responsibilities: make a bridge between Elixir plugins and nvim(load the plugins/translate their interfaces to nvim compatible and return them by a request, allow to use the client API inside plugins). 
Anyway, I've realized that the different projects is overhead here!

But I have another proposals :) (btw, we should close this issue and continue in other tread, it's the last message :smile:  )
I updated the autocomplete script with project-depended autocompletion(inspired by [alchemist-server](https://github.com/tonini/alchemist.el/blob/master/alchemist-server/lib/server.exs)) and bunch of small tweaks: show the first sentence/autoclose the doc-preview window then I go to the next line/etc..
![screenshot](https://cloud.githubusercontent.com/assets/486807/11611935/c79ab77c-9bf5-11e5-97c7-2828b4178b92.png)
![scr](https://cloud.githubusercontent.com/assets/486807/11611937/cc093d1a-9bf5-11e5-984b-2743cebe9c78.png)

I also will add smart refactoring tools to this "script". I write this stuff using usual flow(tests/many levels of abstraction :grin: and so on). And I wanna share this with community. So gists will not work here. But the `mix` already solves this.
The proposal is the adding to `neovim-elixir` framework-like(AKA `fluid` :smile:) responsibilities:
- easy installing 
  `
  mix archive.install ../neovim-elixir-0.1.1.ez, neovim_elixir.install_host
  `
- installing community plugins(as usual packages?)

As a result of this `neovim-elixir` can be able to process usual dynamically loaded `Elixir` scripts and community packaged plugins(a host must be recompiled).

Don't get me wrong, I don't wanna make things complicated. As for me it's just some way to go.
You've already solved most non-trivial problems(thanks!) but there is still room for ​improvement :bowtie:

My priority for all those proposals: 
- HIGH  developer/debug environment
  - the logging
  - integration tests and helpers for a testing
  - docs(how to use NVim interface)
- MID a framework-like feature described above
- very LOW support different elixir versions

**Comment by awetzel** [comment]
Thanks again for your involvement : 

My first attempt to develop a Neovim Host was a "mix archive.install". But there were some issues... the main problem was that all the modules installed as Mix archives are loaded by Mix, so there is a probability that the modules in YOUR project matches the module in the "neovim-elixir" archive and that this collision breaks the neovim host.
I actually encountered the issue myself and then tried to rename all modules in  message_pack with a plugin namespace but it was cumbersome and hacky. Also I thought that the "escriptize" version might includes `erts`  and avoid mismatched version issue, but last issues prove that this is not the case.

Also about the compiled plugin I also tried some architecture to allow that but the drawbacks were:
- the "parsing/compilation" of the plugin module file is not the bottleneck, when you use the Elixir host, because it is insignifiant compared to the VM startup time on first call
- if you do not use a rplugin directory containing files in the targeted langage, then you diverge from the other nvim host plugins and maintenance of the vim plugins in general. What I mean is that a vim package is and has been : "a directory containing files and directory to be merged into your actual vim configs". So if we keep the current system: a new extension elixir plugin is just a "tar" directory or a github repo containing : `/rplugin/elixir/my_new_extension.ex` and the user can use : Pathogen, Vim-Plug, Vundle.vim to install it (with only a `:UpdateRemotePlugin` call compared with standard vim plugins).

But if you achieve it and find an architecture allowing to have a host plugin easier to install and allowing compiled plugin, I would be happy to merge it and to use it :) .

I agree a lot that there is a high priority to add features and docs to make debugging and plugin development easier.

Also the 2 only features missing (compared with alchemist) are:
- Definition lookup (goto definition below cursor)
- Macroexpand.

Macro expand can be done very simply with `Macro.expand/2`, creating a range command which shows the expanded code into the preview window should be a matter of 3 line of codes.
Definition lookup is more difficult to implement but very useful.

**Comment by archseer** [comment]
Hi! To add to the discussion, I've repackaged your vim-elixir fork and added a plugin source for deoplete. https://github.com/archSeer/nvim-elixir I made this fork, because you can now install both the official package, and the neovim additions side by side:

``` vim
Plug 'elixir-lang/vim-elixir'
Plug 'archSeer/nvim-elixir'
```

I'm probably also going to get rid of the tools/ compiled binary and instead add an install script which can be hooked up with Plug to autorun and compile from source.

One change I suggest to add to your tutorial: https://github.com/awetzel/vim-elixir/blob/nvim-rplugin/rplugin/elixir/elixir_plugin.ex#L84 Deoplete and other completion libraries usually call `ElixirComplete(1, "")`, and not `ElixirComplete("1", ""). Because of that, the wrong function definition got called, and [direct integration failed](https://github.com/Shougo/deoplete.nvim/issues/98). I ended up building a [source](https://github.com/archSeer/nvim-elixir/blob/master/rplugin/python3/deoplete/sources/elixir.py), which is recommended over the omnicomplete functionality because it's async.

**Comment by archseer** [comment]
Another thing I noticed with the autocompletion: in a Phoenix project, if I type `Ecto.`, it will show no results, but if I type `Ecto.Query.`, I will get back a list of functions properly. If I try `Ecto.` again, `Ecto.Query.` will show as an autocompletion (note the extra dot), but the other modules will still be missing.

**Comment by awetzel** [comment]
@dm1try  I have added:
- a File logger and a special MIX_ENV: "debug_host" to use it
- the management of archived otp (zipped ".ez") otp app as a plugin

So we keep the ease of use of a `mymodule.ex` rplugin, which is compiled dynamically, but we can use a `myapp.ez` or `myapp-X.Y.Z.ez` rplugin to gain compilation time, otp dependencies start, and multiple file to organize your plugin code.

I have created a project to maintain a "main" rplugin archive plugin : https://github.com/awetzel/nvim-rplugin
@archSeer  and @dm1try  do you think it is a good idea ?

**Comment by awetzel** [comment]
@dm1try do you have the code somewhere of your improvement to the completion function ?

**Comment by awetzel** [comment]
@archSeer , about your repo, vim `Plug` and Deoplete, I actually prefer myself not to use them but your repo is a nice addition for those who does not want to maintain their config, even more if Plug handles the "host" script compilation.  

About your second point, it is actually the behavior of the `IEx` shell completion, since the modules are not necessary loaded and erlang is not in embedded mode, module loading is handled dynamically by the erlang code server, and completion is only for loaded modules.

One solution could be to improve the `MixLoad` command to `:erlang.consult('_build/env/lib/*/ebin/*.app')` and then use `Code.load(module)` for every module in the resulting term. It slows down a lot the mix_load, but since we can do it asynchronously it is ok to do so : 

https://github.com/awetzel/nvim-rplugin/blob/master/lib/nvim_rplugin.ex#L52

**Comment by awetzel** [comment]
something like : 

``` elixir
for f<-Path.wildcard("#{mix_dir}/_build/#{Mix.env}/lib/*/ebin/*.app")),{:ok,[{:application,_,app}]}=:file.consult(to_char_list f),mod<-app[:modules] do
  Code.ensure_loaded(mod)
end
```

https://github.com/awetzel/nvim-rplugin/blob/master/lib/nvim_rplugin.ex#L21-L23

**Comment by archseer** [comment]
Hi @awetzel! Hmm, I'm not sure if it makes sense to maintain two forks of basically the same package (one with deoplete and one without). If you want, I can probably make a PR with my changes; they're optional and won't be loaded unless you're including deoplete as well. Same thing goes for an install script: if you use vim-plug, the script can be automatically ran when installing, but it's useful even without a package manager. The script can be manually ran, which simplifies the entire setup down to a single command.

That said, I do like my package name better :P `<package>.nvim` [seems to be more common](https://github.com/search?q=.nvim&type=Repositories&utf8=%E2%9C%93) than [`-rplugin`](https://github.com/search?utf8=%E2%9C%93&q=-rplugin&type=Repositories&ref=searchresults).

**Comment by awetzel** [comment]
Ok you are right about the single place for the nvim plugin, but the repo I created is an elixir repo to allow compilation of a bigger plugin with multiple files... and does not contain any nvim config files.

So maybe, a proposal (tell me if you have a better idea) : 
3 repos : 
- the Elixir repo containing the elixir Host : https://github.com/awetzel/neovim-elixir
- the Elixir repo containing the elixir rplugin to be archived : https://github.com/awetzel/nvim-rplugin
- The NVim repo containing scripting to compile and install the host and the rplugin, and possibly other vim configurations https://github.com/archSeer/elixir.nvim

Basically, the Elixir.nvim should then : 
- git clone https://github.com/awetzel/neovim-elixir && mix deps.get && MIX_ENV=host && mix escript.build 
- git clone https://github.com/awetzel/nvim-rplugin && mix archive.build
- cp nvim_rplugin-0.0.1.ez nvim_elixir_host respectively into rplugin/elixir and tools

Do you think it is possible ? Do you have a better idea to distinguish elixir host / elixir plugin / nvim config ? maybe mixing repos ?

**Comment by archseer** [comment]
I agree that this repo, `neovim-elixir` with the host implementation should be separate, yes. I think a host (vim API in elixir) + a "client"/user library would work well. Host provides the interface and for developers, the client library contains the autocomplete package, the extra vim commands and support for deoplete and other tools (just like vim-go for example, or vim-ruby).

Because the user package is dependent on the host, I'd provide an `install.sh` that shallow clones (`--depth 1`) the host `neovim-elixir`, compiles it and puts it into the correct spot (<client>/rplugin/\* and <client>/tools)

If the user is installing this manually, he can now just copy the files over into his `.vim` directory, or if he's using a package manager, this is already handled for him.

Package manager installation:

``` vim
Plug '<client>', {do: 'install.sh'}
```

Manual:

``` sh
git clone <client>
cp -R <client> ~/.vim
~/.vim/install.sh
```

**Comment by awetzel** [comment]
Ok I have updated https://github.com/awetzel/nvim-rplugin to improve documentation rendering, to fix your issue and to create a function you can use with `vim-ref` for documentation.

I have created https://github.com/awetzel/elixir.nvim containing the plugin with a little addition: 
https://github.com/awetzel/elixir.nvim/blob/master/autoload/ref/elixir.vim

to have vim-ref available with a nice syntax highlighting in docs. Just press `K` above a module name or a function to get the doc, you can use `K` inside the doc to navigate through it.

**Comment by awetzel** [comment]
@archSeer I added your deoplete source file to my elixir.nvim

**Comment by awetzel** [comment]
I have added background compilation and environment management (both are linked), now Completion, Doc (erlang and elixir), and Evaluation (:ElixirExec) understand the environment: the `alias`, `import`, and even `use` and custom macros, since it actually compile the buffer to get the **ENV** of each module.

https://github.com/awetzel/nvim-rplugin

I have also made configurable the length of the omni completion message and a switch to enable/disable preview window during completion.

**Comment by dm1try** [comment]
@awetzel sorry for the late response. I've installed  [elixit.vim](https://github.com/awetzel/elixir.nvim) and I'm satisfied for now :) Thanks! 
Many useful features were added.
